### ISSUE TITLE
Training with tf.data API gives different results than passing in data directly using Numpy

### ISSUE THREAD

**Issue by oubathAI** [tensorflow/tensorflow#50426]
**System information**
- OS Platform and Distribution : **Linux Ubuntu 18.04**
- TensorFlow installed from  : **source**
- TensorFlow version : **2.5.0**  
- Python version: **3.7.10**
- CUDA/cuDNN version: **CUDA 11.0/cuDNN 8**
- GPU model and memory: **Tesla T4 15GB**

**Describe the current behavior**
Hello,
I have a big difference in the results when I train with the tf.data API compared to when I pass the data directly using Numpy.
Indeed, after 5 epochs, training with tf.data gives the following results 
`Epoch 5/5
690/690 [==============================] - 28s 41ms/step - loss: 1.0746 - accuracy: 0.4408 - val_loss: 1.1136 - val_accuracy: 0.3683`
compared to these results when I train with Numpy:
`Epoch 5/5
690/690 [==============================] - 28s 41ms/step - loss: 0.3204 - accuracy: 0.8693 - val_loss: 5.4233 - val_accuracy: 0.3453`

**Describe the expected behavior**
The expected behavior is to have no difference between tf.data API and Numpy data.

**Standalone code to reproduce the issue**
Here is a link to Google Colab to reproduce the issue: [Google Colab](https://colab.research.google.com/drive/1Ffq2qB7A5jg_ATBUN1qad-3QMDLD2a2p?usp=sharing)

If you can't run Google Colab, here is the code:  [Code to reproduce ](https://gist.github.com/oubathAI/4e20d7b1d7b17b5b4a30a31c3cc8983a)

To download the data, here are the links: 

[train_inputs](https://ec2-spotter.s3.eu-west-1.amazonaws.com/tmp/train_inputs.npy.pkl )
[train_labels](https://ec2-spotter.s3.eu-west-1.amazonaws.com/tmp/train_labels.npy.pkl )
[val_inputs](https://ec2-spotter.s3.eu-west-1.amazonaws.com/tmp/val_inputs.npy.pkl )
[val_labels](https://ec2-spotter.s3.eu-west-1.amazonaws.com/tmp/val_labels.npy.pkl)

**Other info / logs** 
Full training logs using tf.data 
```
Epoch 1/5
690/690 [==============================] - 64s 43ms/step - loss: 1.3643 - accuracy: 0.3942 - val_loss: 1.0925 - val_accuracy: 0.4102
Epoch 2/5
690/690 [==============================] - 28s 40ms/step - loss: 1.0938 - accuracy: 0.3999 - val_loss: 1.1260 - val_accuracy: 0.4157
Epoch 3/5
690/690 [==============================] - 28s 41ms/step - loss: 1.0866 - accuracy: 0.4276 - val_loss: 1.0832 - val_accuracy: 0.4199
Epoch 4/5
690/690 [==============================] - 28s 40ms/step - loss: 1.0871 - accuracy: 0.4186 - val_loss: 1.0825 - val_accuracy: 0.4062
Epoch 5/5
690/690 [==============================] - 28s 41ms/step - loss: 1.0746 - accuracy: 0.4408 - val_loss: 1.1136 - val_accuracy: 0.3683
```

Full training logs using Numpy array
```
Epoch 1/5
690/690 [==============================] - 33s 44ms/step - loss: 1.1418 - accuracy: 0.5239 - val_loss: 1.3560 - val_accuracy: 0.3668
Epoch 2/5
690/690 [==============================] - 28s 41ms/step - loss: 0.6565 - accuracy: 0.7227 - val_loss: 2.4518 - val_accuracy: 0.3578
Epoch 3/5
690/690 [==============================] - 28s 41ms/step - loss: 0.4320 - accuracy: 0.8248 - val_loss: 4.3772 - val_accuracy: 0.3306
Epoch 4/5
690/690 [==============================] - 28s 41ms/step - loss: 0.3603 - accuracy: 0.8544 - val_loss: 4.9844 - val_accuracy: 0.3387
Epoch 5/5
690/690 [==============================] - 28s 41ms/step - loss: 0.3204 - accuracy: 0.8693 - val_loss: 5.4233 - val_accuracy: 0.3453
```

**Comment by bhack** [comment]
Please check https://keras.io/getting_started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development

Also try to print the first element from the dataset and from the numpy array to check that they are exactly the same input in the same order.

**Comment by oubathAI** [comment]
@bhack 
Thanks for your reply.
I tried with `PYTHONHASHSEED=0`  but it didn't change anything.

I build the dataset directly with numpy array `tf.data.Dataset.from_tensor_slices((train_inputs, train_labels))`  ; so the elements are exactly the same when I'm not shuffling the dataset.

**Comment by bhack** [comment]
Have you also set all the seeds as suggested in the FAQ?

`when I'm not shuffling the dataset.`

I think that you need to use the same shuffle method with fixed seeds to exactly reproduce the same batch sequence or to exclude the shuffle if the methods are not exactly the same.

**Comment by oubathAI** [comment]
While I agree that randomness have an impact on the results, what I can't figure out is that every time this randomness gives results where Numpy's way overfit and the tf data API have trouble learning.
Have you looked at the Google Colab? You can run it as many times as you want, and every time the difference between the two ways of training will be the same: The training with Numpy will overfit with the val_loss incrementing, and the loss with TF Dataset will have a hard time decreasing, stuck around 1
I doubt that setting the randomseed will change this behavior, and that's what I'm pointing out in my issue: Why TF Data is not learning while Numpy overfit  everytime I'm trying to train.

**Comment by tilakrayal** [comment]
@sachinprasadhs ,
I was able to reproduce the issue in [v2.4](https://colab.research.google.com/gist/tilakrayal/7ab6b903d4e3c788d000d8984640026c/2-4tfdataset_issue.ipynb),[v2.5](https://colab.research.google.com/gist/tilakrayal/8fb3afb71b18871898b67cc43dac6358/2-5tfdataset_issue.ipynb) and nightly.Please find the gist.

**Comment by sachinprasadhs** [comment]
You need to change the order of  shuffle, batch, cache in your `tf.data` pipeline. 
You should use `cache()` before `shuffle()` ,`batch()` and `prefetch()`, so that your data will be loaded only once instead of each epoch, but data will be shuffled differently during each epoch. 
You should use `shuffle` before `batch`.
I have changed the order and I'm able to see the increase in the accuracy, see the gist [here](https://colab.research.google.com/gist/sachinprasadhs/f5378198aa131a46efab7a960f91d74b/2-5tfdataset_issue.ipynb).

But your model is overfitting, training loss is getting decreased but validation loss is increasing. 
You can rectify using below methods.

- Increase number of samples.
- Shuffle data before splitting into train and validation set.
- Try changing the model layers and other hyper-parameters.
- Use regularization techniques.
- Try reducing batch size.

**Comment by oubathAI** [comment]
Thanks for your reply. I thought I needed to cache at the very last steps of dataset to improve performance.
Thanks for your insighs about the overfitting, I'm trying to mitigate it right now.
Have a nice day!
